### PR TITLE
Сhange type of a tensor with bools

### PR DESCRIPTION
**Pull request by izdeby**
**This is **bc-breaking** change**
Change dtype of a tensor which was created from bool data. 
Old behavior: torch.tensor([True, False]) -> uint8 tensor
Now: torch.tensor([True, False]) -> bool tensor

Tested via tests.

### CODE REVIEW

**Comment by izdeby** [comment]
@gchanan, i think this shows the JIT issue that you were mentioning yesterday. the failing test(test_jit.py::test_torch_tensor) shows that JIT.CompilationUnit is behaving differently. Should i update JITs logic or is it a BC issue and i should add a deprecation warning instead?

File: torch/csrc/utils/tensor_new.cpp [hunk 1]
```diff
@@ -121,8 +121,7 @@ ScalarType infer_scalar_type(PyObject *obj) {
     return ScalarType::Long;
   }
   if (PyBool_Check(obj)) {
-    // TODO: infer Bool when we have Bool ScalarType
-    return ScalarType::Byte;
+    return ScalarType::Bool;
```
Comment (gchanan) [inline torch/csrc/utils/tensor_new.cpp]: all of the non-test changes besides this look correct and can go in (because they don't break backcompat), right?

**Comment by gchanan** [comment]
you should figure out where the JIT logic is first.

File: torch/csrc/jit/register_special_ops.cpp [hunk 2]
```diff
@@ -112,7 +111,7 @@ void storeLastDimension<bool>(
   auto seq_size = obj.size();
   checkSequenceSize(n, dim, seq_size);
   for (int64_t i = 0; i < n; i++) {
-    *(uint8_t*)data = static_cast<uint8_t>(obj[i]);
+    *(bool*)data = static_cast<bool>(obj[i]);
```
Comment (eellison) [inline torch/csrc/jit/register_special_ops.cpp]: I think you need to change   the line here https://github.com/pytorch/pytorch/blob/6fb58f7efd2b615e4103029f6be5244a162e46e2/torch/csrc/jit/register_special_ops.cpp#L293 

from `at::kByte` to the new Dtype

File: torch/csrc/utils/tensor_new.cpp (hunk 1 above)
Comment (izdeby) [reply torch/csrc/utils/tensor_new.cpp]: this and jit changes

**Review by ezyang — APPROVED** [review]

**Pull request closed without merging**